### PR TITLE
Add support for automatically continuing training from earlier runs of a Task (fixes #270)

### LINKED ISSUE #270: Support training continuation for the failed or preempted tasks
It's especially important for pre-emption of spot instances

### PR DESCRIPTION
**Pull request by bhearsum**
Aside from the included tests, I tested this by hand. What I did was:
* Start a `train-backwards` job: https://firefox-ci-tc.services.mozilla.com/tasks/bNP6s4FaRwaU7Bz8gxgJ-Q/runs/0
* Wait for it to checkpoint
* Simulate a spot termination, which was handled correctly, and ended up with 20 artifacts
* The Task automatically reran: https://firefox-ci-tc.services.mozilla.com/tasks/bNP6s4FaRwaU7Bz8gxgJ-Q/runs/1, and picked up the previous artifacts. From the log:
```
[task 2024-05-07T20:54:35.749Z] INFO:root:run_id > 0, attempting to resume training from an earlier run...
[task 2024-05-07T20:54:35.880Z] INFO:root:Run 0 appears to have the artifacts we need! Downloading them...
[task 2024-05-07T20:54:35.880Z] INFO:root:Fetching public/build/config.opustrainer.yml...
[task 2024-05-07T20:54:36.123Z] INFO:root:Fetching public/build/config.opustrainer.yml.state...
[task 2024-05-07T20:54:36.319Z] INFO:root:Fetching public/build/devset.out...
[task 2024-05-07T20:54:36.584Z] INFO:root:Fetching public/build/model.npz...
[task 2024-05-07T20:54:38.649Z] INFO:root:Fetching public/build/model.npz.best-bleu-detok.npz...
[task 2024-05-07T20:54:40.588Z] INFO:root:Fetching public/build/model.npz.best-bleu-detok.npz.decoder.yml...
[task 2024-05-07T20:54:40.770Z] INFO:root:Fetching public/build/model.npz.best-ce-mean-words.npz...
[task 2024-05-07T20:54:42.583Z] INFO:root:Fetching public/build/model.npz.best-ce-mean-words.npz.decoder.yml...
[task 2024-05-07T20:54:42.821Z] INFO:root:Fetching public/build/model.npz.best-chrf.npz...
[task 2024-05-07T20:54:44.517Z] INFO:root:Fetching public/build/model.npz.best-chrf.npz.decoder.yml...
[task 2024-05-07T20:54:44.758Z] INFO:root:Fetching public/build/model.npz.decoder.yml...
[task 2024-05-07T20:54:44.990Z] INFO:root:Fetching public/build/model.npz.optimizer.npz...
[task 2024-05-07T20:54:51.438Z] INFO:root:Fetching public/build/model.npz.progress.yml...
[task 2024-05-07T20:54:51.634Z] INFO:root:Fetching public/build/model.npz.yml...
[task 2024-05-07T20:54:51.834Z] INFO:root:Fetching public/build/opustrainer.log...
[task 2024-05-07T20:54:52.042Z] INFO:root:Fetching public/build/train.log...
[task 2024-05-07T20:54:52.243Z] INFO:root:Fetching public/build/valid.log...
[task 2024-05-07T20:54:52.458Z] INFO:root:Fetching public/build/vocab.spm...
<a bit further down>
[task 2024-05-07T20:57:19.459Z] [2024-05-07 20:57:19] [training] Master parameters and optimizers restored from training checkpoint /home/ubuntu/tasks/task_171511512470744/artifacts/model.npz and /home/ubuntu/tasks/task_171511512470744/artifacts/model.npz.optimizer.npz
```
* I simulated a spot termination on _that_ run quite quickly, and it uploaded the artifacts it had just downloaded. (These get downloaded to the `artifacts` directory when we continue training, so even though _this_ run didn't checkpoint, it essentially just re-uploaded run #0's work).
* Run #2 also started automatically: https://firefox-ci-tc.services.mozilla.com/tasks/bNP6s4FaRwaU7Bz8gxgJ-Q/runs/2, and picked up the artifacts from run #1. Again from the log:
```
[task 2024-05-07T21:04:06.641Z] INFO:root:run_id > 0, attempting to resume training from an earlier run...
[task 2024-05-07T21:04:06.718Z] INFO:root:Run 1 appears to have the artifacts we need! Downloading them...
[task 2024-05-07T21:04:06.718Z] INFO:root:Fetching public/build/config.opustrainer.yml...
[task 2024-05-07T21:04:06.955Z] INFO:root:Fetching public/build/config.opustrainer.yml.state...
[task 2024-05-07T21:04:07.181Z] INFO:root:Fetching public/build/devset.out...
[task 2024-05-07T21:04:07.433Z] INFO:root:Fetching public/build/model.npz...
[task 2024-05-07T21:04:10.476Z] INFO:root:Fetching public/build/model.npz.best-bleu-detok.npz...
[task 2024-05-07T21:04:12.597Z] INFO:root:Fetching public/build/model.npz.best-bleu-detok.npz.decoder.yml...
[task 2024-05-07T21:04:12.812Z] INFO:root:Fetching public/build/model.npz.best-ce-mean-words.npz...
[task 2024-05-07T21:04:15.629Z] INFO:root:Fetching public/build/model.npz.best-ce-mean-words.npz.decoder.yml...
[task 2024-05-07T21:04:15.833Z] INFO:root:Fetching public/build/model.npz.best-chrf.npz...
[task 2024-05-07T21:04:19.250Z] INFO:root:Fetching public/build/model.npz.best-chrf.npz.decoder.yml...
[task 2024-05-07T21:04:19.447Z] INFO:root:Fetching public/build/model.npz.decoder.yml...
[task 2024-05-07T21:04:19.660Z] INFO:root:Fetching public/build/model.npz.optimizer.npz...
[task 2024-05-07T21:04:28.699Z] INFO:root:Fetching public/build/model.npz.progress.yml...
[task 2024-05-07T21:04:28.941Z] INFO:root:Fetching public/build/model.npz.yml...
[task 2024-05-07T21:04:29.166Z] INFO:root:Fetching public/build/opustrainer.log...
[task 2024-05-07T21:04:29.399Z] INFO:root:Fetching public/build/train.log...
[task 2024-05-07T21:04:29.609Z] INFO:root:Fetching public/build/valid.log...
[task 2024-05-07T21:04:29.842Z] INFO:root:Fetching public/build/vocab.spm...
<a bit further down>
[task 2024-05-07T21:06:50.280Z] [2024-05-07 21:06:50] [training] Master parameters and optimizers restored from training checkpoint /home/ubuntu/tasks/task_171511569156440/artifacts/model.npz and /home/ubuntu/tasks/task_171511569156440/artifacts/model.npz.optimizer.npz
```

(I ended up canceling run 2 to avoid wasting resources.)

Additional interpretation of the logs and artifacts is welcome; I'm also happy to do more simulated test runs if that's useful.

Landing this depends on [an update to the GPU images](https://phabricator.services.mozilla.com/D209840) that includes the latest spot termintation handling code. I expect this to land in the next day.

### CODE REVIEW

File: taskcluster/scripts/pipeline/train_taskcluster.py [hunk 1]
```diff
@@ -1,14 +1,101 @@
...
+import logging
+import os
 import os.path
+import requests
 import subprocess
 import sys
 
 TRAINING_SCRIPT = os.path.join(os.path.dirname(__file__), "train-taskcluster.sh")
+CONTINUATION_ARTIFACTS = {
+    "config.opustrainer.yml",
```
Comment (bhearsum) [inline taskcluster/scripts/pipeline/train_taskcluster.py:12]: @eu9ene - I originally used the list from https://github.com/mozilla/firefox-translations-training/blob/99015e13d50cb2b3222d916fefc71e78505639fe/taskcluster/translations_taskgraph/transforms/training_continuation.py#L5 for this, but found it didn't work due to `final` models being present. Please sanity check this list.

And tangentially related: should we update the training continuation artifact list to expect opustrainer files?

Comment (eu9ene) [reply taskcluster/scripts/pipeline/train_taskcluster.py:12]: Yes, to continue training from the same place OpusTrainer needs `config.opustrainer.yml.state` (this is a default path for the state file). I see that it already outputs this to the artifacts.

Ideally, we should modify the train.sh not to rebuild the config again if it's already present but this code is deterministic so it should work.

File: tests/test_train_taskcluster.py
Comment (gabrielBusta) [inline tests/test_train_taskcluster.py:1]: Yay for tests!

**Review by gabrielBusta — APPROVED** [review]

File: taskcluster/scripts/pipeline/train_taskcluster.py [hunk 2]
```diff
@@ -1,14 +1,101 @@
...
+                logging.info(f"Fetching {artifact['name']}...")
+
+                r = requests.get(
+                    f"{root_url}/api/queue/v1/task/{task_id}/runs/{prev_run_id}/artifacts/{artifact['name']}",
+                    stream=True,
+                )
+                r.raise_for_status()
+
+                with open(os.path.join(model_dir, out_name), "wb+") as fd:
+                    for chunk in r.iter_content(chunk_size=8192):
```
[the remark this answers is not in the document]
Comment (bhearsum) [reply taskcluster/scripts/pipeline/train_taskcluster.py:85]: Honestly - it was just what I copy/pasted from StackOverflow, and it seemed like a reasonable amount of data to write at one time.

File: taskcluster/scripts/pipeline/train_taskcluster.py [hunk 3]
```diff
@@ -1,14 +1,101 @@
...
+
+    script_args = list(args)
+    task_id = os.environ["TASK_ID"]
+    run_id = int(os.environ["RUN_ID"])
+    root_url = os.environ["TASKCLUSTER_ROOT_URL"]
+    # Must line up with where model_dir is in `train-taskcluster.sh` while that script
+    # still exists.
+    model_dir = script_args[6]
+    pretrained_model_mode = None
+    if len(args) >= 11:
```
[the remark this answers is not in the document]
Comment (bhearsum) [reply taskcluster/scripts/pipeline/train_taskcluster.py:44]: Good call.

[the remark this answers is not in the document]
Comment (eu9ene) [reply taskcluster/scripts/pipeline/train_taskcluster.py]: I recently added another argument

File: taskcluster/scripts/pipeline/train_taskcluster.py [hunk 4]
```diff
@@ -1,14 +1,101 @@
...
+                continue
+
+            for artifact in resp.json()["artifacts"]:
+                # Skip Taskcluster logs - we only care about artifacts that the training tools create.
+                if artifact["name"].startswith("public/log"):
+                    continue
+                out_name = os.path.basename(artifact["name"])
+                logging.info(f"Fetching {artifact['name']}...")
+
+                r = requests.get(
```
[the remark this answers is not in the document]
Comment (bhearsum) [reply taskcluster/scripts/pipeline/train_taskcluster.py:78]: I'm going to forego this for the moment, mainly because getting the tests working is not trivial, and it's only a minor benefit.

File: taskcluster/scripts/pipeline/train_taskcluster.py [hunk 5]
```diff
@@ -1,14 +1,101 @@
...
+                    f"{root_url}/api/queue/v1/task/{task_id}/runs/{prev_run_id}/artifacts/{artifact['name']}",
+                    stream=True,
+                )
+                r.raise_for_status()
```
Comment (eu9ene) [inline taskcluster/scripts/pipeline/train_taskcluster.py:92]: I wonder what happens if there's an exception in this code, say from data downloading. Will the task just be retried? Do we want to implement any retries right here?

File: tests/test_train_taskcluster.py [hunk 6]
```diff
@@ -76,8 +80,261 @@
...
+                        "model-type",
+                        "training-type",
+                        "src",
+                        "trg",
+                        "train-set-prefix",
+                        "valid-set-prefix",
+                        model_dir,
+                        "best-model-metric",
+                        "alignents",
+                        "seed",
```
Comment (eu9ene) [inline tests/test_train_taskcluster.py:283]: make sure to sync with the main as there are more args now

**Review by eu9ene — APPROVED** [review]
I'm approving after reviewing the last [training run](https://firefox-ci-tc.services.mozilla.com/tasks/d5jrxdfRThyt7BtSbUYgJw/runs/3). W&B dashboards and Marian training logs look correct. There are some comments to address about the number of arguments though.

File: tests/test_train_taskcluster.py (hunk 6 above)
Comment (bhearsum) [reply tests/test_train_taskcluster.py:283]: Thanks, will do!

**Comment by bhearsum** [comment]
I've kicked off two test runs to validate the latest changes:
* https://firefox-ci-tc.services.mozilla.com/tasks/O9QxLh3FQ2uE29lb1y257w is the same as previous runs, where I wait for a checkpoint, simulate a spot termination, and verify that it continues. I've done this once so far, and will do it again after the next checkpoint
* https://firefox-ci-tc.services.mozilla.com/tasks/D2fDAVu2Qge5ZHoMT9W8Lg is a run with [this additional change to force a download error](https://github.com/mozilla/firefox-translations-training/commit/5ff8f13bc205546006d090305be57f8171efa719) to prove that the new retry logic works (it does!).

File: taskcluster/scripts/pipeline/train_taskcluster.py [hunk 7]
```diff
@@ -1,14 +1,125 @@
...
+# The argument number where pretrained model mode is expected.
+# This is 1-indexed, not 0-indexed, so it should line up with the argument
+# number this is fetched in in train-taskcluster.sh
+PRETRAINED_MODEL_MODE_ARG_NUMBER = 12
+# Nothing special about 17...just a number plucked out of thin air that
+# should be distinct enough to retry on.
+DOWNLOAD_ERROR_EXIT_CODE = 17
 
 
 def main(args):
```
Comment (gregtatum) [inline taskcluster/scripts/pipeline/train_taskcluster.py:44]: Thought: I guess ultimately we'll peel back all the layers to this onion and merge all of these separate scripts into a single training script that can just handle all of this, and write it all in python. I filed: #607 for this.

File: tests/test_train_taskcluster.py [hunk 8]
```diff
@@ -76,8 +80,262 @@
...
+        "storageType": "s3",
+        "name": "public/logs/live.log",
+        "expires": "2035-04-23T16:22:13.477Z",
+        "contentType": "text/plain; charset=utf-8",
+    },
+]
+
+
+@pytest.mark.parametrize(
+    "current_run_id,resumable_run_id,reasons_created,run_artifacts,orig_pretrained_model_mode,expected_pretrained_model_mode",
```
Comment (gregtatum) [inline tests/test_train_taskcluster.py:135]: Thought (no action need): This style of parameterization is hard to connect the parameters with their names. I'm not sure if there is a way to label them easier with a TypedDict or something. I do like the fact that they have `id`s on them.

File: tests/test_train_taskcluster.py [hunk 9]
```diff
@@ -76,8 +80,262 @@
...
+                    )
+                else:
+                    resp.status_code = 400
+                    resp._content = f"train_taskcluster.py made a call to an unexpected URL: {url}"
+
+                return resp
+
+            tt_mock["requests"].get = mock.Mock(wraps=fake_get)
+
+            with tempfile.TemporaryDirectory() as model_dir:
```
Comment (gregtatum) [inline tests/test_train_taskcluster.py:271]: Suggestion: We have the `DataDir` fixture that puts these directories in the same place. This makes it easier debug failing tests.

```
data_dir = DataDir("test_train_taskcluster")
data_dir.path
```

**Review by gregtatum — APPROVED** [review]
Thanks! Looks good to me.

File: taskcluster/scripts/pipeline/train_taskcluster.py (hunk 7 above)
Comment (bhearsum) [reply taskcluster/scripts/pipeline/train_taskcluster.py:44]: Yeah. I'm not proud of introducing this extra layer of indirection, but I do intend to follow up when there's more time.

File: tests/test_train_taskcluster.py (hunk 8 above)
Comment (bhearsum) [reply tests/test_train_taskcluster.py:135]: I agree. I'm not aware of any way to do this with [pytest params](https://docs.pytest.org/en/latest/example/parametrize.html) :(. It looks like https://github.com/pytest-dev/pytest/issues/9216 talks about this a bit.

File: tests/test_train_taskcluster.py (hunk 9 above)
Comment (bhearsum) [reply tests/test_train_taskcluster.py:271]: That's handy! Let's use it here too.

**Comment by bhearsum** [comment]
#605 appears to be breaking finetune-student in the PR triggered tasks :(

> I've kicked off two test runs to validate the latest changes:
> 
>     * https://firefox-ci-tc.services.mozilla.com/tasks/O9QxLh3FQ2uE29lb1y257w is the same as previous runs, where I wait for a checkpoint, simulate a spot termination, and verify that it continues. I've done this once so far, and will do it again after the next checkpoint

The run is now complete, and worked fine with the two simulated spot terminations.
